File: aquecimento intrif 2025/P/P.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <map>
#include <queue>

// Estrutura para representar uma estrada na lista de adjacência
struct Edge {
    int to;
    int id;
    int color;
};

// Estrutura para guardar os detalhes de uma estrada para consulta
struct Road {
    int u, v, color;
};

// Variáveis globais para acesso na função recursiva
int N, M, K;
std::vector<std::vector<Edge>> adj;
std::vector<Road> roads;

// Função de busca recursiva (DFS)
void find_tour(int start_node, int current_node, std::vector<int>& path, std::vector<bool>& used_roads) {
    // Caso base: o caminho está completo
    if (path.size() == M) {
        if (current_node == start_node) {
            int first_road_color = roads[path[0]].color;
            int last_road_color = roads[path.back()].color;

            if (first_road_color != last_road_color) {
                // Solução encontrada! Imprime e encerra o programa.
                std::cout << start_node << std::endl;
                for (size_t i = 0; i < path.size(); ++i) {
                    std::cout << path[i] << (i == path.size() - 1 ? "" : " ");
                }
                std::cout << std::endl;
                exit(0);
            }
        }
        return;
    }

    int last_road_color = roads[path.back()].color;

    // Passo recursivo: tenta todas as arestas a partir do nó atual
    for (const auto& edge : adj[current_node]) {
        if (!used_roads[edge.id]) {
            if (edge.color != last_road_color) {
                used_roads[edge.id] = true;
                path.push_back(edge.id);

                find_tour(start_node, edge.to, path, used_roads);

                // Backtrack
                path.pop_back();
                used_roads[edge.id] = false;
            }
        }
    }
}

bool solve() {
    if (!(std::cin >> N >> M >> K)) {
        return false;
    }

    adj.assign(N + 1, std::vector<Edge>());
    roads.assign(M + 1, Road());
    std::vector<int> degree(N + 1, 0);

    for (int i = 1; i <= M; ++i) {
        int u, v, c;
        std::cin >> u >> v >> c;
        roads[i] = {u, v, c};
        adj[u].push_back({v, i, c});
        adj[v].push_back({u, i, c});
        degree[u]++;
        degree[v]++;
    }

    // --- PRÉ-VERIFICAÇÕES ---
    std::vector<int> potential_starts;
    int nodes_with_edges = 0;
    for(int i = 1; i <= N; ++i) {
        if (degree[i] > 0) {
            nodes_with_edges++;
            potential_starts.push_back(i);
        }
    }
    
    if (M > 0 && nodes_with_edges == 0) {
        std::cout << -1 << std::endl;
        return true;
    }
     if (M == 0) {
        std::cout << 1 << std::endl << std::endl;
        return true;
    }

    // 1. Verificação de Conectividade (BFS)
    std::queue<int> q;
    q.push(potential_starts[0]);
    std::vector<bool> visited_component(N + 1, false);
    visited_component[potential_starts[0]] = true;
    int component_size = 1;
    while(!q.empty()){
        int u = q.front();
        q.pop();
        for(const auto& edge : adj[u]){
            if(!visited_component[edge.to]){
                visited_component[edge.to] = true;
                q.push(edge.to);
                component_size++;
            }
        }
    }
    if(component_size != nodes_with_edges){
        std::cout << -1 << std::endl;
        return true;
    }

    // 2. Verificação de Grau Par
    for (int i = 1; i <= N; ++i) {
        if (degree[i] % 2 != 0) {
            std::cout << -1 << std::endl;
            return true;
        }
    }

    // 3. Verificação de Distribuição de Cores
    for (int i = 1; i <= N; ++i) {
        if (degree[i] > 0) {
            std::map<int, int> color_counts;
            for (const auto& edge : adj[i]) {
                color_counts[edge.color]++;
            }
            for (const auto& pair : color_counts) {
                if (pair.second > degree[i] / 2) {
                    std::cout << -1 << std::endl;
                    return true;
                }
            }
        }
    }

    // --- BUSCA ROBUSTA ---
    for (int start_node : potential_starts) {
        std::vector<bool> used_roads(M + 1, false);
        for (const auto& edge : adj[start_node]) {
            std::vector<int> path;
            path.push_back(edge.id);
            used_roads[edge.id] = true;

            find_tour(start_node, edge.to, path, used_roads);

            used_roads[edge.id] = false; // Backtrack da escolha da primeira aresta
        }
    }

    // Se todos os pontos de partida falharam
    std::cout << -1 << std::endl;
    return true;
}

int main() {
    // Otimização de I/O para C++
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    // Loop para múltiplos casos de teste, se houver
    solve();

    return 0;
}

File: aquecimento intrif 2025/P/P.py
import sys

# Aumenta o limite de recursão para caminhos de busca profundos (até M estradas)
sys.setrecursionlimit(2050)

def find_tour(start_node, current_node, path, visited_roads, M, roads, adj):
    """
    Função recursiva (DFS) que busca por um passeio válido.
    Retorna True se uma solução for encontrada e impressa, False caso contrário.
    """
    # Caso base: O caminho incluiu todas as estradas
    if len(path) == M:
        # Verifica se o passeio terminou na cidade inicial
        if current_node == start_node:
            first_road_color = roads[path[0]][2]
            last_road_color = roads[path[-1]][2]
            
            # Verifica se a primeira e a última estrada têm cores diferentes
            if first_road_color != last_road_color:
                print(start_node)
                print(*path)
                return True
        return False

    # Passo recursivo: Explora as próximas estradas possíveis
    last_road_color = roads[path[-1]][2]
    
    for edge in adj[current_node]:
        road_id = edge['id']
        road_color = edge['color']
        
        if not visited_roads[road_id]:
            if road_color != last_road_color:
                visited_roads[road_id] = True
                path.append(road_id)
                
                if find_tour(start_node, edge['neighbor'], path, visited_roads, M, roads, adj):
                    return True
                
                path.pop()
                visited_roads[road_id] = False
    return False

def solve():
    """
    Função principal para ler um único caso de teste, verificar condições e encontrar o passeio.
    """
    try:
        line = sys.stdin.readline()
        if not line.strip():
            return False
        N, M, K = map(int, line.split())
    except (IOError, ValueError):
        return False

    adj = [[] for _ in range(N + 1)]
    roads = [None] * (M + 1)
    degrees = [0] * (N + 1)

    for road_id in range(1, M + 1):
        u, v, c = map(int, sys.stdin.readline().split())
        roads[road_id] = (u, v, c)
        adj[u].append({'neighbor': v, 'id': road_id, 'color': c})
        adj[v].append({'neighbor': u, 'id': road_id, 'color': c})
        degrees[u] += 1
        degrees[v] += 1

    # --- Verificações de Pré-requisitos (executadas uma vez) ---
    first_node_with_edges = -1
    nodes_with_edges = 0
    potential_starts = []
    for i in range(1, N + 1):
        if degrees[i] > 0:
            nodes_with_edges += 1
            potential_starts.append(i)
            if first_node_with_edges == -1:
                first_node_with_edges = i

    if M > 0 and first_node_with_edges == -1:
        print(-1)
        return True

    if M == 0:
        print(1 if N >= 1 else -1)
        if N >= 1: print()
        return True

    q = [first_node_with_edges]
    visited_component = {first_node_with_edges}
    head = 0
    while head < len(q):
        u = q[head]; head += 1
        for edge in adj[u]:
            v = edge['neighbor']
            if v not in visited_component:
                visited_component.add(v)
                q.append(v)
    
    if len(visited_component) != nodes_with_edges:
        print(-1)
        return True

    for i in range(1, N + 1):
        if degrees[i] % 2 != 0:
            print(-1)
            return True

    for i in range(1, N + 1):
        if degrees[i] > 0:
            color_counts = {}
            for edge in adj[i]:
                color = edge['color']
                color_counts[color] = color_counts.get(color, 0) + 1
            
            for color in color_counts:
                if color_counts[color] > degrees[i] // 2:
                    print(-1)
                    return True

    # --- Busca Robusta com Múltiplos Pontos de Partida ---
    # Itera sobre todas as cidades possíveis como ponto de partida.
    for start_node in potential_starts:
        visited_roads = [False] * (M + 1)
        
        # Para cada cidade inicial, tenta começar com cada uma de suas estradas.
        for edge in adj[start_node]:
            path = [edge['id']]
            visited_roads[edge['id']] = True
            
            if find_tour(start_node, edge['neighbor'], path, visited_roads, M, roads, adj):
                # Se uma solução for encontrada e impressa, o caso de teste termina.
                return True
            
            # Backtrack para tentar a próxima estrada inicial
            visited_roads[edge['id']] = False

    # Se, após tentar todas as cidades como ponto de partida, nenhuma solução for encontrada.
    print(-1)
    return True

if __name__ == "__main__":
    while solve():
        pass